python/ray/air/execution/resources/resource_manager.py

```diff
@@ -0,0 +1,123 @@
+    if resources are not needed anymore.
+
+    Thus, resources managed by the resource manager have three states:
+
+    1. Requested but not available (ready). The resource request is received,
+       but the resources are not available for use of remote Ray objects, yet.
+    2. Requested and ready. The resources can be acquired and subsequently used by
+       remote Ray objects.
+    3. Acquired. These resources have been acquired by a caller.
+
+    The flow is as follows:
+
+    .. code-block:: python
+
+        # Create resource manager
+        resource_manager = ResourceManager()
+
+        # Create resource request
+        resource_request = ResourceRequest([{"CPU": 4}])
+
+        # Pass to resource manager
+        resource_manager.request_resources(resource_request)
+
+        # Wait until ready
+        while not resource_manager.has_resources_ready(resource_request):
+            time.sleep(1)
+
+        # Once ready, acquire resources
+        acquired_resource = resource_manager.acquire_resources(resource_request)
+
+        # Bind to remote task or actor
+        annotated_remote_fn = acquired_resource.annotate_remote_objects(
+            [remote_fn])
+
+        # Run remote function. This will use the acquired resources
+        ray.get(annotated_remote_fn.remote())
+
+        # After using the resources, free
+        resource_manager.free_resources(annotated_resources)
+
+    """
+
+    def request_resources(self, resource_request: ResourceRequest):
+        """Request resources, e.g. schedule placement group."""
+        raise NotImplementedError
+
+    def cancel_resource_request(self, resource_request: ResourceRequest):
+        """Cancel resource request, e.g. remove placement group."""
+        raise NotImplementedError
+
+    def has_resources_ready(self, resource_request: ResourceRequest) -> bool:
+        """Returns True if resources for the given request are available"""
+        raise NotImplementedError
+
+    def acquire_resources(
+        self, resource_request: ResourceRequest
+    ) -> Optional[AcquiredResource]:
+        """Acquire resources. Returns None if resources are not available."""
+        raise NotImplementedError
+
+    def free_resources(self, acquired_resource: AcquiredResource):
+        """Free resources from usage."""
+        raise NotImplementedError
+
+    def get_resource_futures(self) -> List[ray.ObjectRef]:
+        """Return futures for resources to await.
+
+        Depending on the backend, we use resource futures to determine availability
+        of resources (e.g. placement groups). In this case, they can be awaited
+        externally by an outer control loop.
+
+        When a resource future resolved, you may want to call ``update_state()``
+        to force the resource manager to update its internal state, which will
+        make the resources available to be used on the next call.
+        """
+        return []
+
+    def update_state(self):
```

Comment:
One suggestion is to make this much more specific as to what state is being reconciled (requested -> ready resources).

```suggestion
    def reconcile_requested_resource_state(self):
```